### PR TITLE
[WIP] core: add support for clip skip to SDXL

### PR DESCRIPTION
**Pull request by sayakpaul**
Following:

https://github.com/huggingface/diffusers/pull/4901/

Example code:

```python
from diffusers import DiffusionPipeline
from diffusers.utils import make_image_grid
import torch

pipe = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"


skips = [None, 1, 2, 3, 4]
final_images = []

for clip_skip in skips:
    generator = torch.manual_seed(245)
    image = pipe(
        prompt, generator=generator, num_inference_steps=25, clip_skip=clip_skip
    ).images[0]
    final_images.append(image)

make_image_grid(final_images, 1, len(final_images)).save("skip_results.png")
```

Pretty cool results:

skip_results

### CODE REVIEW

**Comment by bghira** [comment]
nit: ticket title

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 1]
```diff
@@ -305,14 +306,15 @@ def encode_prompt(
...
+                prompt_embeds = text_encoder(text_input_ids.to(device), output_hidden_states=True)
 
                 # We are only ALWAYS interested in the pooled output of the final text encoder
                 pooled_prompt_embeds = prompt_embeds[0]
-                prompt_embeds = prompt_embeds.hidden_states[-2]
+                if clip_skip is None:
+                    prompt_embeds = prompt_embeds.hidden_states[-2]
+                else:
+                    # "2" because SDXL always indexes from the penultimate layer.
+                    prompt_embeds = prompt_embeds.hidden_states[-(clip_skip + 2)]
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:317]: Yes think that's how we should do it here. The other option would be to set `clip_skip` to 1 by default but I think that's not a good idea. 

So good as is, thanks!

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sayakpaul** [comment]
@patrickvonplaten added thr `clip_skip` argument to all the supported pipelines in SDXL. PTAL and then we can merge.

**Review by patrickvonplaten — APPROVED** [review]
Looks great! Thanks for applying this to all pipelines

**Pull request merged**